### ISSUE TITLE
Unable to resolve the template type U in call to method loophp\collection\Contract\Operation\Associateable<(int|string),Symfony\Component\Uid\Uuid>::associate()

### ISSUE THREAD

**Issue by rela589n** [loophp/collection#349]
## Steps required to reproduce the problem

Have the collection of `Uuid` object (e.g. `Collection<array-key,Uuid>`). And try to call `associate()` in order to reindex the keys:

```php
$uuidCollection->associate(static fn ($key, Uuid $value): string => $value->toRfc4122())
```

## Expected Result

No warning when the `$callbackForValues` is not given

## Actual Result

The warning from phpstan:

```
  392    Unable to resolve the template type U in call to method loophp\collection\Contract\Operation\Associateable<(int|string),Symfony\Component\Uid\Uuid>::associate()                                                                                                    
         🪪  argument.templateType                                                                                                                                                                                                                                           
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type                                                                                                                                                                              
         ✏️  
```

There's a warning, making us stumble and think how to work this around.

**Comment by drupol** [comment]
Thank you for the report, please provide a PR if you can and we'll continue from there.

**Comment by rela589n** [comment]
BTW, do you think if there any use-cases of `associate()` with two parameters (both `$callbackForKeys` and `$callbackForValues`) that could not be implemented with `associate()` + `$callbackForKeys` and just `map`?

**Comment by drupol** [comment]
I have no clue about it, I have no way to know which method is used by this library users, sadly.
However, `associate` let to modify `keys` and `values` at the same time. If you're only modifies `values`, then I would advise to use `map` obviously.

**Comment by rela589n** [comment]
@drupol , what do you think about `keyMap(TKey, T, iterable<TKey, T>)` method?

**Comment by rela589n** [comment]
Regarding the original `associate` method, I guess the following return type would have been more correct, since there are four different outcomes depending on whether `$callbackForKeys` and `$callbackForValues` are given:

```diff
===================================================================
@@ -20,10 +20,14 @@
      * @template UKey
      * @template U
      *
-     * @param callable(TKey, T, iterable<TKey, T>): UKey $callbackForKeys
-     * @param callable(T, TKey, iterable<TKey, T>): U $callbackForValues
+     * @param null|callable(TKey, T, iterable<TKey, T>): UKey $callbackForKeys
+     * @param null|callable(T, TKey, iterable<TKey, T>): U $callbackForValues
      *
-     * @return Collection<UKey, U>
+     * @psalm-return (
+     *  $callbackForKeys is null
+     *  ? ($callbackForValues is null ? Collection<TKey, T> : Collection<TKey, U>)
+     *  : ($callbackForValues is null ? Collection<UKey, T> : Collection<UKey, U>)
+     * )
      */
     public function associate(?callable $callbackForKeys = null, ?callable $callbackForValues = null): Collection;
 }

```

**Comment by drupol** [comment]
Sounds good ;)

Let's do a PR to fix this too !

**Comment by rela589n** [comment]
I have tried to sort out the issue by writing down the correct conditional return types, and ... :clock10: ...  I found myself stuck in a static analysis rat hole I couldn't get out from. The problem is that phpstan/psalm doesn't think that `CollectionDecorator` has correct returned value type (that's both the first problem I encountered and the one which I left off). It says that returned type must be covariant with the interface :man_shrugging: 

<img src="https://github.com/user-attachments/assets/404ad3df-2a4b-42e3-8744-480f6ed92e00" width="196">


Finally, I think the root problem here is not even with static analysis. It is trying to put two things into a single `associate` method, since it results in four different combinations of the outcome. 

So, there are four different outcomes:
(1) when both callbacks are absent, - in this scenario, neither keys nor values are modified, - usually it makes no sense to call it that way and if we had two different methods, it would correspond to not calling either of them;
(2,3) when only one callback is provided - if there's keys callback, the keys are modified, otherwise values callback modifies values; if we had two different methods, this would correspond to one particular method call of the one particular respective method;
(4) when both callbacks are provided - both keys and values are modified (if we had two different methods, it would correspond to two lines of code each calling the respective particular method).

Therefore, in my opinion, the right way would be to have two separate methods - one specifically to modify keys, and another specifically to modify values. Having both of them combined in `associate` method violates SRP and makes it a nightmare of static analysis.
